gapir/client/client.go

```diff
@@ -102,33 +140,152 @@ func (client *Client) Connect(ctx context.Context, device bind.Device, abi *devi
 	if err != nil {
-		return nil, log.Err(ctx, err, "Background connection error")
+		return nil, log.Err(ctx, err, "Error in startReplayCommunicationHandler")
 	}
 
-	client.clientInfos[key] = clientInfo{
-		deviceConnectionInfo: *newDeviceConnectionInfo,
-		connection:           connection,
-		device:               device,
-		arch:                 abi.Architecture,
-		abi:                  abi,
-		bgConnection:         bgConnection}
+	client.replayers[key] = replayerInfo
 	return &key, nil
 }
 
-func (client *Client) makeBackgroundConnection(ctx context.Context, device bind.Device, conn gapir.Connection) (*backgroundConnection, error) {
-	bgc := &backgroundConnection{conn: conn, OS: device.Instance().GetConfiguration().GetOS()}
+// removeConnection closes and removes a GAPIR instance.
+func (client *Client) removeConnection(ctx context.Context, key ReplayerKey) {
+	client.mutex.Lock()
+	defer client.mutex.Unlock()
+
+	if replayer, ok := client.replayers[key]; ok {
+		replayer.closeConnection(ctx)
+		delete(client.replayers, key)
+	}
+}
+
+// reconnect removes the replayer before re-creating it.
+func (client *Client) reconnect(ctx context.Context, replayer *replayerInfo) {
+	key := ReplayerKey{device: replayer.device, arch: replayer.abi.GetArchitecture()}
+	client.removeConnection(ctx, key)
+	client.Connect(ctx, replayer.device, replayer.abi)
+}
+
+// heartbeat regularly sends a ping to a replayer, and restarts it when it fails to reply.
+func (client *Client) heartbeat(ctx context.Context, replayer *replayerInfo) {
+	for {
+		select {
+		case <-task.ShouldStop(ctx):
+			return
+		case <-time.After(heartbeatInterval):
+			err := replayer.ping(ctx)
+			if err != nil {
+				log.E(ctx, "Error sending keep-alive ping. Error: %v", err)
+				client.reconnect(ctx, replayer)
+				return
+			}
+		}
+	}
+}
+
+// getActiveReplayer returns the replayer identified by key only if this
+// replayer has an active connection to a GAPIR instance.
+func (client *Client) getActiveReplayer(ctx context.Context, key *ReplayerKey) (*replayerInfo, error) {
+	replayer, found := client.replayers[*key]
+	if !found {
+		return nil, log.Errf(ctx, nil, "Cannot find replayer for this key: %v", key)
+	}
+
+	if replayer.rpcClient == nil || replayer.conn == nil || replayer.rpcStream == nil {
+		return nil, log.Err(ctx, nil, "Replayer has no active connection")
+	}
+
+	return replayer, nil
+}
+
+// BeginReplay sends a replay request to the replayer identified by key.
+func (client *Client) BeginReplay(ctx context.Context, key *ReplayerKey, payload string, dependent string) error {
+	ctx = log.Enter(ctx, "Starting replay on gapir device")
+	replayerInfo, err := client.getActiveReplayer(ctx, key)
+	if err != nil {
+		return err
+	}
+
+	idReq := replaysrv.ReplayRequest{
+		Req: &replaysrv.ReplayRequest_Replay{
+			Replay: &replaysrv.Replay{
+				ReplayId:    payload,
+				DependentId: dependent,
+			},
+		},
+	}
+	err = replayerInfo.rpcStream.Send(&idReq)
```

Comment:
Even though I like removing files and reducing complexity, we may still want to separate rpc layer from this for both separation of concerns and testing.